### ISSUE TITLE
Add support for curving line2d

### ISSUE THREAD

**Issue by iapps207** [godotengine/godot-proposals#1106]
**Describe the project you are working on:**
Mobile game
**Describe the problem or limitation you are having in your project:**
I want to curve my line2d node
**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
It's an innovation I call line2d.add_curve_point
Basically add_curve_point method on line2d nodes 
**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
It creates curve and all you have to choose is whether to have an inward or outward curve but defaults to outward
**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
No
**Is there a reason why this should be core and not an add-on in the asset library?:**
Also part of gdscript

**Comment by jonbonazza** [comment]
But if it curves, it's no longer a line. 🤔

**Comment by CodeZombie** [comment]
Godot already has a Curve2D object that does what you want.

**Comment by Xrayez** [comment]
There's no node which could draw the `Curve2D` resource, so I think the proposal is partially relevant. But yeah I think you could still draw a curve with the `Line2D` similar to:

```gdscript
$line.points = curve.get_baked_points()
```

I'm also not sure whether the `Line2D` name is confusing. It's more like a `Polyline2D`, and not `Segment2D`.

Also, the `Path2D` node *does* have a `curve` property which can be set via editor. It's currently not the case for `Line2D`, so perhaps we're talking about consistency here.

**Comment by iapps207** [comment]
> There's no node which could draw the `Curve2D` resource, so I think the proposal is partially relevant. But yeah I think you could still draw a curve with the `Line2D` similar to:
> 
> ```gdscript
> $line.points = curve.get_baked_points()
> ```
> 
> I'm also not sure whether the `Line2D` name is confusing. It's more like a `Polyline2D`, and not `Segment2D`.
> 
> Also, the `Path2D` node _does_ have a `curve` property which can be set via editor. It's currently not the case for `Line2D`, so perhaps we're talking about consistency here.

Yes we are talking about consistency and also there is no code for implementing close in Line2d like:
$line2d.close_points

Which means we have to use set the end point to the starting point

**Comment by iapps207** [comment]
@Calinou This proposal is useful for drawing complex shapes like my tooltip UI in the my previous issue

**Comment by Xrayez** [comment]
> Which means we have to use set the end point to the starting point

Yeah, there are other places where this kind of feature would be useful in the drawing API, see godotengine/godot#34923 and my following comments regarding this.

**Comment by Xrayez** [comment]
This is why I suggested the `Polyline2D` name, because it can be "closed", that's not necessarily the case for `Line2D`. But likely the "closed" property is something which is an inherent part of the existing `Polygon2D` class, but there's no way to draw a polygon non-filled with the color (so that only the boundary is visible), so I dunno.

**Comment by jordanlis** [comment]
Hey, so what about this evolution ? I would like to draw a curve in the game. This could be great to support curve for line2D like it is already done with path2D.

Here, there's no curve edition
![image](https://user-images.githubusercontent.com/68544949/132584714-f5937e91-79aa-42bd-8805-83f078b4b74c.png)

And here, there is
![image](https://user-images.githubusercontent.com/68544949/132584830-da210cd4-f7a1-486e-8dca-3c102b57c617.png)

It could be usefull to draw simple curves in the game like it is already possible for path2D

**Comment by Zireael07** [comment]
I think this issue should be renamed to 'convert Line2D to path2D', since that would achieve the goal the OP is after without bloating line2d and turning it into a dupe of path2d

**Comment by jordanlis** [comment]
> 
> 
> I think this issue should be renamed to 'convert Line2D to path2D', since that would achieve the goal the OP is after without bloating line2d and turning it into a dupe of path2d

I don't think so, no. The goal is to curve 2Dline, not use path2D. I hope everything is clear ;p

**Comment by Calinou** [comment]
> I don't think so, no. The goal is to curve 2Dline, not use path2D. I hope everything is clear ;p

We can't afford making Line2D too complex. Otherwise, it will become impossible to maintain. We also prefer simpler composable solutions to large, "one size fits all" solutions.

**Comment by jordanlis** [comment]
too bad, it could have been a good feature.

**Comment by IoneGod** [comment]
Well then you could create CurveLine2D node that should be more modular I guess 🤔

**Comment by MagdielM** [comment]
> Well then you could create CurveLine2D node that should be more modular I guess 🤔

I'd also consider a `ShapeRenderer2D` node that takes shapes and curves provided by child nodes and displays them in the viewport, almost like a `CanvasLayer`. This would avoid bloating existing shape nodes too much, although some changes will probably be necessary, given that according to the docs:

> The exact drawing method is specific for each shape and cannot be configured.

Honestly, I might come up with a proposal for this later.

**Comment by IoneGod** [comment]
> > Well then you could create CurveLine2D node that should be more modular I guess 🤔
> 
> I'd also consider a `ShapeRenderer2D` node that takes shapes and curves provided by child nodes and displays them in the viewport, almost like a `CanvasLayer`. This would avoid bloating existing shape nodes too much, although some changes will probably be necessary, given that according to the docs:
> 
> > The exact drawing method is specific for each shape and cannot be configured.
> 
> Honestly, I might come up with a proposal for this later.

I like this ide, is this possible or is there a possible future where this becomes unmaintainable? @devs

**Comment by SquiggelSquirrel** [comment]
Somewhat relevant to https://github.com/godotengine/godot-proposals/issues/4330

**Comment by Calinou** [comment]
@CsloudX Please don't bump issues without contributing significant new information. Use the :+1: reaction button on the first post instead.

**Comment by peterhoglund** [comment]
> Not sure there is a push, but it's pretty trivial to implement to your spec if you need it using `Node2D`, `_draw`, and a `Curve2D` resource.

Is it possible to `_draw` a texture or gradient the way you can with Line2D? If not, I think adding bezier to Line2D is not only a QoL addition but something expected from a "vector style drawing" tool.

**Comment by kcfresh53** [comment]
I made a plugin that may be able to solve this issue in the meantime. Here is a link to the [repo](https://github.com/kcfresh53/bezier-line2d).

**Comment by renevanderark** [comment]
> I made a plugin that may be able to solve this issue in the meantime. Here is a link to the repo.

Yup, later @kcfresh53 teamed up with me to work on https://github.com/Teaching-myself-Godot/ez-curved-lines-2d for a while (which was really cool)

It ships a custom node that does Bézier curve management, you can basically do most of what svg can as nodes with it. Yes, too much solve everything, but it could be made more composed by breaking it up into parts.

I get a lot of response along the lines of: I wish this were in the core/I wish I could get the shockwave flash experience back.

We're inching closer, but in-engine would be the real deal, right?
